sdkit/utils/file_utils.py

```diff
@@ -34,23 +36,39 @@ def save_images(images: list, dir_path: str, file_name='image', output_format='J
         path = os.path.join(dir_path, actual_file_name)
         img.save(f'{path}.{output_format.lower()}', quality=output_quality)
 
-def save_dicts(entries: list, dir_path: str, file_name='data', output_format='txt'):
+def save_dicts(entries: list, dir_path: str, file_name='data', output_format='txt', file_format=''):
     '''
     * entries: a list of dictionaries
     * dir_path: the directory path where the files will be saved
     * file_name: the file name to save. Can be a string or a function.
         if a string, the actual file name will be `{file_name}_{index}`.
         if a function, the callback function will be passed the `index` (int),
           and the returned value will be used as the actual file name. e.g `def fn(i): return 'foo' + i`
-    * output_format: 'txt' or 'json'
+    * output_format: 'txt', 'json', or 'embed'
+        if 'embed', the metadata will be embedded in PNG files in tEXt chunks, and as EXIF UserComment for JPEG files
     '''
-    if output_format.lower() not in ['json', 'txt']: return
     if dir_path is None: return
     os.makedirs(dir_path, exist_ok=True)
 
     for i, metadata in enumerate(entries):
         actual_file_name = file_name(i) if callable(file_name) else f'{file_name}_{i}'
         path = os.path.join(dir_path, actual_file_name)
+
+        if output_format.lower() == 'embed' and file_format.lower() == 'png':
+            targetImage = Image.open(f'{path}.{file_format.lower()}')
+            embedded_metadata = PngInfo()
+            for key, val in metadata.items():
+                embedded_metadata.add_text(key, str(val))
+            targetImage.save(f'{path}.{file_format.lower()}', pnginfo=embedded_metadata)
+            return
+        elif output_format.lower() == 'embed' and file_format.lower() == 'jpeg':
+            targetImage = Image.open(f'{path}.{file_format.lower()}')
+            user_comment = json.dumps(metadata)
+            exif_data = targetImage.getexif()
+            exif_data[0x9286] = user_comment
+            targetImage.save(f'{path}.{file_format.lower()}', exif=exif_data)
+            return
```

Comment:
entries can be a list of dictionaries. Calling return here would only save the first item of that list.

Reply:
Validate using parallel renders. Only the first file will have the metadata.